tests/unit/test_library_smtp.py

```diff
@@ -0,0 +1,272 @@
+    act: update the relation data.
+    assert: the relation data is updated.
+    """
+    harness = Harness(SmtpProviderCharm, meta=PROVIDER_METADATA)
+    harness.begin()
+    harness.set_leader(True)
+    harness.add_relation("smtp-legacy", "smtp-provider")
+    relation = harness.model.get_relation("smtp-legacy")
+    smtp_data = smtp.SmtpRelationData(
+        host="example.smtp",
+        port=25,
+        auth_type="plain",
+        transport_security="tls",
+    )
+    harness.charm.smtp_legacy.update_relation_data(relation, smtp_data)
+    data = relation.data[harness.model.app]
+    assert data["host"] == smtp_data.host
+    assert data["port"] == str(smtp_data.port)
+    assert data["auth_type"] == smtp_data.auth_type
+    assert data["transport_security"] == smtp_data.transport_security
+
+
+def test_smtp_relation_data_to_relation_data():
+    """
+    arrange: instantiate a SmtpRelationData object.
+    act: obtain the relation representation.
+    assert: the relation representation is correct.
+    """
+    smtp_data = smtp.SmtpRelationData(
+        host="example.smtp",
+        port=25,
+        user="example_user",
+        password="somepassword",  # nosec
+        password_id="someid",
+        auth_type="plain",
+        transport_security="tls",
+        domain="domain",
+    )
+    relation_data = smtp_data.to_relation_data()
+    expected_relation_data = {
+        "host": "example.smtp",
+        "port": "25",
+        "user": "example_user",
+        "password": "somepassword",  # nosec
+        "password_id": "someid",
+        "auth_type": "plain",
+        "transport_security": "tls",
+        "domain": "domain",
+    }
+    assert relation_data == expected_relation_data
+
+
+def test_legacy_requirer_charm_does_not_emit_event_id_when_no_data():
+    """
+    arrange: set up a charm with no relation data to be populated.
+    act: trigger a relation changed event.
+    assert: no events are emitted.
+    """
+    harness = Harness(SmtpRequirerCharm, meta=REQUIRER_METADATA)
+    harness.begin()
+    harness.set_leader(True)
+    harness.add_relation("smtp-legacy", "smtp-provider")
+    relation = harness.charm.framework.model.get_relation("smtp-legacy", 0)
+    harness.charm.on.smtp_legacy_relation_changed.emit(relation)
+    assert len(harness.charm.events) == 0
+
+
+def test_requirer_charm_does_not_emit_event_id_when_no_data():
+    """
+    arrange: set up a charm with no relation data to be populated.
+    act: trigger a relation changed event.
+    assert: no events are emitted.
+    """
+    harness = Harness(SmtpRequirerCharm, meta=REQUIRER_METADATA)
+    harness.begin()
+    harness.set_leader(True)
+    harness.add_relation("smtp", "smtp-provider")
+    relation = harness.charm.framework.model.get_relation("smtp", 0)
+    harness.charm.on.smtp_legacy_relation_changed.emit(relation)
+    assert len(harness.charm.events) == 0
+
+
+@pytest.mark.parametrize("is_leader", [True, False])
```

Comment:
Do we need this? Test charms simply count events. They don't care for unit leadership.

Reply:
This is actually a regression test

Reply:
Regression test of what exactly? There's nothing depending on the leadership in the implementation, so I don't understand the point of checking this.

Reply:
There was at some point

Reply:
Right, but not anymore. So why keep it? :)

Reply:
The reason to keep it is to ensure that behaviour continues to be the case.

Reply:
I see. And why is that important?
Actually, this makes me wonder... Since you're writing to `app` data in the relation data bag, perhaps it would actually make sense if only the leader could do that? WDYT?

Reply:
You can see the discussion [here](https://github.com/canonical/saml-integrator-operator/pull/32#discussion_r1293308008). Let me know if you need more info

Reply:
No, I'm good. I see the full context now. Thanks. Please close it.